operator/internal/lifecycle/client.go

```diff
@@ -0,0 +1,387 @@
+	// since this is a widespread deletion, we can delete even stateful sets
+	resources, err := r.listAllOwnedResources(ctx, owner, true)
+	if err != nil {
+		return false, err
+	}
+
+	alive := []client.Object{}
+	for _, o := range resources {
+		if o.GetDeletionTimestamp() == nil {
+			alive = append(alive, o)
+		}
+	}
+
+	// attempt to delete as many resources in one pass as we can
+	errs := []error{}
+	for _, resource := range alive {
+		if err := r.client.Delete(ctx, resource); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return len(alive) > 0, errors.Join(errs...)
+}
+
+// listResources lists resources of a specific type and object, returning them as an array.
+func (r *ResourceClient[T, U]) listResources(ctx context.Context, object client.Object, opts ...client.ListOption) ([]client.Object, error) {
+	kind, err := getGroupVersionKind(r.client.Scheme(), object)
+	if err != nil {
+		return nil, err
+	}
+
+	olist, err := r.client.Scheme().New(schema.GroupVersionKind{
+		Group:   kind.Group,
+		Version: kind.Version,
+		Kind:    kind.Kind + "List",
+	})
+	if err != nil {
+		return nil, fmt.Errorf("initializing list: %w", err)
+	}
+	list, ok := olist.(client.ObjectList)
+	if !ok {
+		return nil, fmt.Errorf("invalid object list type: %T", object)
+	}
+
+	if err := r.client.List(ctx, list, opts...); err != nil {
+		return nil, fmt.Errorf("listing resources: %w", err)
+	}
+
+	converted := []client.Object{}
+	items := reflect.ValueOf(list).Elem().FieldByName("Items")
+	for i := 0; i < items.Len(); i++ {
+		item := items.Index(i).Addr().Interface().(client.Object)
+		item.GetObjectKind().SetGroupVersionKind(*kind)
+		converted = append(converted, item)
+	}
+
+	return sortCreation(converted), nil
+}
+
+// listAllOwnedResources lists all resources owned by a given cluster, optionally including node pools.
+func (r *ResourceClient[T, U]) listAllOwnedResources(ctx context.Context, owner U, includeNodePools bool) ([]client.Object, error) {
+	resources := []client.Object{}
+	for _, resourceType := range append(r.simpleResourceRenderer.WatchedResourceTypes(), &appsv1.StatefulSet{}) {
+		matching, err := r.listResources(ctx, resourceType, client.MatchingLabels(r.ownershipResolver.GetOwnerLabels(owner)))
+		if err != nil {
+			return nil, err
+		}
+		filtered := []client.Object{}
+		for i := range matching {
+			// special case the node pools
+			if includeNodePools || !r.nodePoolRenderer.IsNodePool(matching[i]) {
+				filtered = append(filtered, matching[i])
+			}
+		}
+		resources = append(resources, filtered...)
+	}
+	return resources, nil
+}
+
+// patchOwnedResource applies a patch to a resource owned by the cluster.
+func (r *ResourceClient[T, U]) patchOwnedResource(ctx context.Context, owner U, object client.Object, extraLabels ...map[string]string) error {
+	if err := r.normalize(object, owner, extraLabels...); err != nil {
+		return err
+	}
+	return r.client.Patch(ctx, object, client.Apply, defaultFieldOwner, client.ForceOwnership)
```

Comment:
I _strongly_ prefer that we not use a global value for field owner here. It should be a member on the struct or be inherited from `r.client`.